### PR TITLE
Eigen Conversion - Part 2

### PR DESCRIPTION
**Pull request by mrivi**
Same as #232 but rebased on current master.

Conversion Quaternion/Euler isn't yet consistent with the previously used getRPY method
1) Correct
```bash
Q(w,x,y,z) 0.001000 0.008836 0.024367 0.999582 
R=
  -0.999516 -0.00156857   0.0177143
 0.00242986   -0.998485   0.0486966
  0.0176168   0.0487319    0.998656
Eigen yaw 3.14002 pitch  0.0487512 roll -0.0176444
matrix 
-0.999842 -0.001569 0.017717 
0.002430 -0.998810 0.048704 2
0.017620 0.048740 0.998656 
getRPY sol1 3.13916 -0.0176206 0.0487667
getRPY sol2 -0.00243063 3.15921 -3.09283
```

2) Not Correct
```bash
Q(w,x,y,z) -0.003000 0.008836 0.024367 0.999582 
R=
 -0.999516 0.00642808  0.0175193
-0.0055668  -0.998485  0.0487672
 0.0178117  0.0486612   0.998656
Eigen yaw 0.00643775 pitch  3.09291 roll 3.12375
matrix 
-0.999826 0.006429 0.017522 
-0.005568 -0.998794 0.048775 
0.017814 0.048669 0.998656 
getRPY sol1 -3.13602 -0.0178154 0.0486957
getRPY sol2 0.00556857 3.15941 -3.0929
```

### CODE REVIEW

**Comment by mrivi** [comment]
I went trought the code of the tf and eigen libraries to make sure the quaternion to euler angles transformation are correct. 

The rotation matrix definition are the same. In Eigen
```bash
R = 
1 - (2yy + 2zz), 2xy - 2zw,        2zx + 2yw
2xy + 2zw,       1 - (2xx + 2zz),  2zy - 2xw
2zx - 2yw,       2zy + 2xw,        1 - (2xx + 2yy)
```

while in the tf package https://github.com/ros/geometry/blob/d718a1f4ff206e5d9196fbc1156293929583bed2/tf/include/tf/LinearMath/Matrix3x3.h#L148
```bash
R =
1 - (2yy + 2zz) / d , (2xy - 2zw) / d,        (2zx + 2yw) / d
(2xy + 2zw) / d,      1 - (2xx + 2zz) / d,    (2zy - 2xw) / d
(2zx - 2yw) / d,      (2zy + 2xw) / d,        1 - (2xx + 2yy) / d
```
where `d = xx + yy + zz + ww`
If the quaternion length, d, is ~1 , the rotation matrices are the same. 

If the tf package `getRPY`  ( ( https://github.com/ros/geometry/blob/d718a1f4ff206e5d9196fbc1156293929583bed2/tf/include/tf/LinearMath/Matrix3x3.h#L294-L360 ) first the pitch, then roll, then yaw angles are calculated.
Threfore the equivalent in eigen (https://github.com/libigl/eigen/blob/1f05f51517ec4fd91eed711e0f89e97a7c028c0e/Eigen/src/Geometry/EulerAngles.h#L37) is EulerAngles(1, 0, 2)

However the transformations are not yet consistent between the two libraries.

File: local_planner/src/nodes/common.cpp [hunk 1]
```diff
@@ -158,18 +149,61 @@ Eigen::Vector3f toEigen(const pcl::PointXYZ& p) {
   return ev3;
 }
 
+Eigen::Quaternionf toQuaternion(const geometry_msgs::Quaternion& gmq) {
```
Comment (jkflying) [inline local_planner/src/nodes/common.cpp]: Rather call this `toEigen`

File: local_planner/include/local_planner/local_planner.h [hunk 2]
```diff
@@ -64,6 +60,7 @@ class LocalPlanner {
 
   float velocity_mod_;
```
Comment (jkflying) [inline local_planner/include/local_planner/local_planner.h]: Maybe this can be deleted now

File: local_planner/src/nodes/local_planner_node.cpp [hunk 3]
```diff
@@ -122,7 +120,7 @@ LocalPlannerNode::~LocalPlannerNode() { delete server_; }
 
 void LocalPlannerNode::readParams() {
   // Parameter from launch file
-  auto goal = local_planner_->getGoal();
+  auto goal = toPoint(local_planner_->getGoal());
```
Comment (jkflying) [inline local_planner/src/nodes/local_planner_node.cpp]: Rather do the toPoint down below when it is actually going into the goal_msg_, and up here use an Eigen type

File: local_planner/src/nodes/local_planner_node.cpp [hunk 4]
```diff
@@ -749,15 +718,12 @@ void LocalPlannerNode::publishGround() {
...
   int beta_z = floor(
-      (atan2(x - drone_pos.pose.position.x, y - drone_pos.pose.position.y) *
+      (atan2(x - drone_pos.x(), y - drone_pos.y()) *
        180.0 / M_PI));  //(-180. +180]
-  int beta_e = floor((atan((z - drone_pos.pose.position.z) /
-                           sqrt((x - drone_pos.pose.position.x) *
-                                    (x - drone_pos.pose.position.x) +
-                                (y - drone_pos.pose.position.y) *
-                                    (y - drone_pos.pose.position.y))) *
+  int beta_e = floor((atan((z - drone_pos.z()) / sqrt((x - drone_pos.x()) *
```
Comment (jkflying) [inline local_planner/src/nodes/local_planner_node.cpp]: `sqrt((x - drone_pos.x()) * (x - drone_pos.x()) + (y - drone_pos.y()) * (y - drone_pos.y()))`

can be replaced with

`(Eigen::Vector2f(x,y) - drone_pos.topRows<2>()).norm()` 

which I think is a lot more readable

File: local_planner/src/nodes/waypoint_generator.cpp [hunk 5]
```diff
@@ -89,25 +86,25 @@ void WaypointGenerator::calculateWaypoint() {
...
                                     bool stay, bool is_airborne) {
-  pose_ = act_pose;
-  curr_vel_ = vel;
-  goal_ = toEigen(goal.pose.position);
-  curr_yaw_ = static_cast<float>(tf::getYaw(pose_.pose.orientation));
+
+  position_ = act_pose;
+  velocity_ = vel;
+  goal_ = goal;
+  Eigen::Vector3f euler = q.toRotationMatrix().eulerAngles(1, 0, 2);
```
Comment (jkflying) [inline local_planner/src/nodes/waypoint_generator.cpp]: You can get the yaw inline just using the `.z()` on the output ie. 

`curr_yaw_ = q.toRotationMatrix().eulerAngles(1, 0, 2).z();`

File: local_planner/src/nodes/local_planner_node.cpp (hunk 3 above)
Comment (mrivi) [reply local_planner/src/nodes/local_planner_node.cpp]: goal needs to be a geometry_msgs type because of the launch params assignments in the lines below.

**Comment by mrivi** [comment]
@jkflying there is still some issue with the quaternion/euler conversion. See my previous comment.

**Comment by mrivi** [comment]
@jkflying I reverted to the tf package for the quaternion to euler conversion because it will take me more time to figure out what's wrong.

Sitl Logs: https://review.px4.io/plot_app?log=d03de169-0f6c-4cec-95c7-de02e2231dfd
https://review.px4.io/plot_app?log=16a81856-1681-4798-bf1b-cbc2a7bb6edd

**Comment by jkflying** [comment]
I've SITL tested this, it is behaving the same as current master.

Interestingly, outside of the node itself, `toEigen()` is not used except for the PCL pointcloud types, and `toPoint()` is not used at all! Nice work!

**Pull request merged**